crates/subspace-farmer/src/plot.rs

```diff
@@ -376,3 +304,201 @@ impl WeakPlot {
+            .map_err(PlotError::PlotOpen)?;
+
+        let piece_count = Arc::new(AtomicU64::new(plot_size / PIECE_SIZE as u64));
+
+        let piece_offset_to_index_file = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .create(true)
+            .open(base_directory.as_ref().join("plot-offset-to-index.bin"))
+            .map_err(PlotError::OffsetDbOpen)?;
+
+        let piece_index_to_offset_db =
+            DB::open_default(base_directory.as_ref().join("plot-index-to-offset"))
+                .map(IndexToOffsetDB::from)
+                .map_err(PlotError::IndexDbOpen)?;
+
+        Ok(Self {
+            plot,
+            piece_index_to_offset_db,
+            piece_offset_to_index_file,
+            piece_count,
+        })
+    }
+
+    pub fn read_encoding(
+        &mut self,
+        piece_index: PieceIndex,
+        mut buffer: impl AsMut<[u8]>,
+    ) -> io::Result<()> {
+        let offset = self
+            .piece_index_to_offset_db
+            .get(piece_index)?
+            .ok_or_else(|| io::Error::other(format!("Piece with {piece_index} not found")))?;
+        self.plot.seek(SeekFrom::Start(offset))?;
+        self.plot.read_exact(buffer.as_mut())
+    }
+
+    fn get_piece_index(&mut self, offset: PieceOffset) -> io::Result<PieceIndex> {
+        let mut buf = [0; 8];
+        self.piece_offset_to_index_file.seek(SeekFrom::Start(
+            offset * std::mem::size_of::<PieceIndex>() as u64,
+        ))?;
+        self.piece_offset_to_index_file.read_exact(&mut buf)?;
+        Ok(u64::from_be_bytes(buf))
+    }
+
+    fn put_piece_index(&mut self, offset: PieceOffset, index: PieceIndex) -> io::Result<()> {
+        self.piece_offset_to_index_file.seek(SeekFrom::Start(
+            offset * std::mem::size_of::<PieceIndex>() as u64,
+        ))?;
+        self.piece_offset_to_index_file
+            .write_all(&index.to_be_bytes())
+    }
+
+    pub fn do_plot(
+        mut self,
+        any_requests_receiver: mpsc::Receiver<()>,
+        read_requests_receiver: mpsc::Receiver<ReadRequests>,
+        write_requests_receiver: mpsc::Receiver<WriteRequests>,
+    ) {
+        let mut did_nothing = true;
+        loop {
+            if did_nothing {
+                // Wait for stuff to come in
+                if any_requests_receiver.recv().is_err() {
+                    break;
+                }
+            }
+
+            did_nothing = true;
+
+            // Process as many read requests as there is
+            while let Ok(read_request) = read_requests_receiver.try_recv() {
+                did_nothing = false;
+
+                match read_request {
+                    ReadRequests::ReadEncoding {
+                        index,
+                        result_sender,
+                    } => {
+                        let mut buffer = Piece::default();
+                        let result = self.read_encoding(index, &mut buffer).map(|()| buffer);
```

Comment:
Hmm, why pass in the buffer?

```suggestion
                        let result = self.read_encoding(index)?;
```

Reply:
Because we might want to allocate buffer for `N` pieces and call method `N` times for `N` slices of same buffer.